### PR TITLE
Drop support for PHP < 7.1 and Symfony < 3.4

### PR DESCRIPTION
**Pull request by alcaeus**
As per discussion in Slack and #885, this PR removes support for PHP < 7.1 (as they are no longer supported) as well as Symfony < 3.4 (since they are in security support only). This will target the 1.11 minor release.

The remaining build failure is due to a deprecation which isn't ignored in the Symfony 3.4 build. I have no idea which dependency combination causes that, so if anyone would have a few pointers for me, they'd be very much appreciated.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -31,34 +29,37 @@ script:
...
-    - php: 5.6
-      env: SYMFONY_VERSION="2.8.*"
-
-    - php: 7.0
-      env: SYMFONY_VERSION="3.2.*"
+    # Test against latest Symfony 3.4 stable
+    - php: 7.1
+      env: SYMFONY_VERSION="3.4.*"
 
+    # Test against latest Symfony 3.4 dev version
```
Comment (greg0ire) [inline .travis.yml]: Will there ever be a 3.4-RC version again? Should it be `stability=dev` below?

File: composer.json [hunk 2]
```diff
@@ -24,28 +24,29 @@
...
-        "symfony/yaml": "~2.7|~3.0|~4.0",
-        "symfony/validator": "~2.7|~3.0|~4.0",
-        "symfony/property-info": "~2.8|~3.0|~4.0",
-        "symfony/phpunit-bridge": "~2.7|~3.0|~4.0",
-        "twig/twig": "~1.26|~2.0",
+        "doctrine/orm": "^2.6",
+        "symfony/yaml": "^3.4|^4.1",
+        "symfony/validator": "^3.4|^4.1",
+        "symfony/property-info": "^3.4|^4.1",
+        "symfony/phpunit-bridge": "^3.4|^4.1",
```
Comment (greg0ire) [inline composer.json]: This one is special, you can require `^4.2` for it, there's no harm.

File: .travis.yml (hunk 1 above)
Comment (alcaeus) [reply .travis.yml]: I've updated it to `dev` below.

File: composer.json (hunk 2 above)
Comment (alcaeus) [reply composer.json]: I keep forgetting about that, thanks!

**Comment by alcaeus** [comment]
@doctrine/team-symfony-integration Anyone know where the deprecation in https://travis-ci.org/doctrine/DoctrineBundle/jobs/466345293 is coming from? It only happens in the Symfony 3.4 build where we lock all our explicit dependencies to Symfony 3.4 (but allow newer versions of other dependencies). I can't quite figure it out what's triggering it...

**Comment by greg0ire** [comment]
@alcaeus you could try getting a stack trace by using `SYMFONY_DEPRECATIONS_HELPER=/MessageSelector/`

**Comment by alcaeus** [comment]
Thanks for the hint! I traced it to the service container but I'm not sure what the root cause is:
```
The "Symfony\Component\Translation\MessageSelector" class is deprecated since Symfony 4.2, use IdentityTranslator instead.
Stack trace:
#0 vendor/symfony/translation/MessageSelector.php(14): trigger_error('The "Symfony\\Co...', 16384)
#1 vendor/composer/ClassLoader.php(444): include('/Users/abraun/C...')
#2 vendor/composer/ClassLoader.php(322): Composer\Autoload\includeFile('/Users/abraun/C...')
#3 [internal function]: Composer\Autoload\ClassLoader->loadClass('Symfony\\Compone...')
#4 /private/var/folders/rn/1wpg9zfn4gbd84dl26xzbfzm0000gn/T/sf_kernel_15f0720cc98d408ed5540a9748774e29/cache/test/ContainerB38ugye/sf_kernel_15f0720cc98d408ed5540a9748774e29TestDebugProjectContainer.php(567): spl_autoload_call('Symfony\\Compone...')
#5 /private/var/folders/rn/1wpg9zfn4gbd84dl26xzbfzm0000gn/T/sf_kernel_15f0720cc98d408ed5540a9748774e29/cache/test/ContainerB38ugye/getValidator_BuilderService.php(21): ContainerB38ugye\sf_kernel_15f0720cc98d408ed5540a9748774e29TestDebugProjectContainer->getTranslator_DefaultService()
#6 /private/var/folders/rn/1wpg9zfn4gbd84dl26xzbfzm0000gn/T/sf_kernel_15f0720cc98d408ed5540a9748774e29/cache/test/ContainerB38ugye/sf_kernel_15f0720cc98d408ed5540a9748774e29TestDebugProjectContainer.php(346): require('/private/var/fo...')
#7 /private/var/folders/rn/1wpg9zfn4gbd84dl26xzbfzm0000gn/T/sf_kernel_15f0720cc98d408ed5540a9748774e29/cache/test/ContainerB38ugye/getValidator_Mapping_CacheWarmerService.php(8): ContainerB38ugye\sf_kernel_15f0720cc98d408ed5540a9748774e29TestDebugProjectContainer->load('getValidator_Bu...')
#8 /private/var/folders/rn/1wpg9zfn4gbd84dl26xzbfzm0000gn/T/sf_kernel_15f0720cc98d408ed5540a9748774e29/cache/test/ContainerB38ugye/sf_kernel_15f0720cc98d408ed5540a9748774e29TestDebugProjectContainer.php(346): require('/private/var/fo...')
#9 /private/var/folders/rn/1wpg9zfn4gbd84dl26xzbfzm0000gn/T/sf_kernel_15f0720cc98d408ed5540a9748774e29/cache/test/ContainerB38ugye/getCacheWarmerService.php(9): ContainerB38ugye\sf_kernel_15f0720cc98d408ed5540a9748774e29TestDebugProjectContainer->load('getValidator_Ma...')
#10 vendor/symfony/http-kernel/CacheWarmer/CacheWarmerAggregate.php(43): ContainerB38ugye\sf_kernel_15f0720cc98d408ed5540a9748774e29TestDebugProjectContainer->{closure}()
#11 vendor/symfony/http-kernel/Kernel.php(554): Symfony\Component\HttpKernel\CacheWarmer\CacheWarmerAggregate->warmUp('/private/var/fo...')
#12 vendor/symfony/http-kernel/Kernel.php(123): Symfony\Component\HttpKernel\Kernel->initializeContainer()
#13 Tests/Command/ImportMappingDoctrineCommandTest.php(28): Symfony\Component\HttpKernel\Kernel->boot()
#14 vendor/phpunit/phpunit/src/Framework/TestCase.php(725): Doctrine\Bundle\DoctrineBundle\Tests\Command\ImportMappingDoctrineCommandTest->setup()
#15 {main}
```

File: DependencyInjection/DoctrineExtension.php [hunk 3]
```diff
@@ -65,7 +66,7 @@ public function load(array $configs, ContainerBuilder $container)
             throw new LogicException('Configuring the ORM layer requires to configure the DBAL layer as well.');
         }
 
-        if (! class_exists('Doctrine\ORM\Version')) {
+        if (! class_exists(Version::class)) {
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply DependencyInjection/DoctrineExtension.php]: It works until ORM 3.0, so I'd keep with it for now and only change it when we release version 2.0 of the bundle for ORM 3.0.

File: Tests/ProfilerTest.php [hunk 4]
```diff
@@ -73,9 +73,7 @@ public function testRender()
         // This is only needed for WebProfilerBundle=3.2, remove when support for it is dropped
         $requestCollector = new RequestDataCollector();
```
Comment (ostrolucky) [inline Tests/ProfilerTest.php]: I remember this one, you can remove whole block actually. No need to add $requestCollector to profile

Comment (alcaeus) [reply Tests/ProfilerTest.php]: Thanks, I removed this 👍

**Comment by alcaeus** [comment]
Thanks @ostrolucky for finding the cause of the deprecation. For now, I changed the deprecation helper config to report all deprecations without failing the build - this solves the issue regardless of whether it's fixed upstream.

File: Repository/ContainerRepositoryFactory.php [hunk 5]
```diff
@@ -30,7 +29,7 @@ public function __construct(ContainerInterface $container = null)
     {
         // When DoctrineBundle requires Symfony 3.3+, this can be removed
         // and the $container argument can become required.
-        if ($container === null && class_exists(ServiceLocatorTagPass::class)) {
+        if ($container === null) {
             throw new InvalidArgumentException(sprintf('The first argument of %s::__construct() is required on Symfony 3.3 or higher.', self::class));
```
Comment (ostrolucky) [inline Repository/ContainerRepositoryFactory.php]: Shouldn't we just make it non-nullable in constructor, since it throws exception otherwise anyway?

**Comment by ostrolucky** [comment]
How about setting `SYMFONY_DEPRECATIONS_HELPER=1` instead? That would ignore the current deprecation, but alert us when count increases, which might be useful

File: .travis.yml [hunk 6]
```diff
@@ -31,34 +29,37 @@ script:
 
 jobs:
   include:
-    - php: 5.5
-      env: deps=low SYMFONY_DEPRECATIONS_HELPER=weak
+    # Tests the lowest set of depedencies
```
Comment (greg0ire) [inline .travis.yml]:
```suggestion
    # Tests the lowest set of dependencies
```

File: Repository/ContainerRepositoryFactory.php (hunk 5 above)
Comment (alcaeus) [reply Repository/ContainerRepositoryFactory.php]: Unfortunately, that would change the behaviour from an exception to a fatal error, which I would consider a BC break. We'll do that for 2.0.

**Comment by alcaeus** [comment]
@ostrolucky 

> Can you please remove argument for HttpKernelExtension in setUp method of ProfilerTest? It seems it's no longer needed. I think it was for older versions too

Will do 👍 

> How about setting `SYMFONY_DEPRECATIONS_HELPER=1` instead? That would ignore the current deprecation, but alert us when count increases, which might be useful

I'd rather not have the build fail on deprecations in older builds to prevent regressions like the one before (where a newer version of a peer dependency triggers a deprecation). The tests against the latest version should fail when deprecations arise, so we should spot them soon enough 👍

File: .travis.yml [hunk 7]
```diff
@@ -31,34 +29,37 @@ script:
...
-      env: stability=dev SYMFONY_DEPRECATIONS_HELPER=weak SYMFONY_VERSION="4.1.*"
+      env: stability=dev SYMFONY_DEPRECATIONS_HELPER=999999 SYMFONY_VERSION="4.1.*"
       install:
         - composer require --dev "symfony/messenger:4.1.*" --no-update
         - travis_retry composer update -n --prefer-dist
 
+    # Test against latest Symfony 4.2 dev version
     - php: 7.1
-      env: stability=dev SYMFONY_DEPRECATIONS_HELPER=weak SYMFONY_VERSION="4.2.*"
+      env: stability=dev SYMFONY_DEPRECATIONS_HELPER=999999 SYMFONY_VERSION="4.2.*"
```
Comment (nicolas-grekas) [inline .travis.yml]: Shouldn't that one fail on deprecations?
Running with the latest version of Symfony should always be deprecation.
That will help you spot where the bundle is behind and fix.

Comment (alcaeus) [reply .travis.yml]: This one just tests against an upcoming 4.2 release which should not contain any new deprecations anyways. We always test against the latest stable version of Symfony using the regular build jobs.

It may be worth adding a separate job that tests against the master branch of Symfony (similar to how we currently test with `'minimum-stability': 'alpha'`) to report upcoming deprecations, but then I'd also have that build allow failures so as to not break unrelated PRs. What do you think?

Comment (nicolas-grekas) [reply .travis.yml]: > This one just tests against an upcoming 4.2 release which should not contain any new deprecations anyways

so better ensure it by failing if this is not verified? this would prevent merging a PR in the bundle that'd use a deprecated feature.

> adding a separate job that tests against the master branch

that'd make sense yes, failing on deprecations but in allowed failure. It would help identify which parts of the bundle need to be upgraded in preparation for the next version of Symfony.

Comment (alcaeus) [reply .travis.yml]: That sounds good. Will update the pipeline 👍

Comment (alcaeus) [reply .travis.yml]: Done!

File: .travis.yml (hunk 6 above)
Comment (alcaeus) [reply .travis.yml]: Done, thanks!

**Comment by alcaeus** [comment]
@ostrolucky apparently that argument is still required, the build against the lowest set of dependencies fails without it.

**Comment by ostrolucky** [comment]
Yeah, it installed TwigBridge 2.8 heh

File: DependencyInjection/DoctrineExtension.php [hunk 8]
```diff
@@ -168,13 +162,13 @@ protected function loadDbalConnection($name, array $connection, ContainerBuilder
         }
 
         // event manager
-        $container->setDefinition(sprintf('doctrine.dbal.%s_connection.event_manager', $name), new $definitionClassname('doctrine.dbal.connection.event_manager'));
+        $container->setDefinition(sprintf('doctrine.dbal.%s_connection.event_manager', $name), new ChildDefinition('doctrine.dbal.connection.event_manager'));
```
Comment (greg0ire) [inline DependencyInjection/DoctrineExtension.php]: Should we take this as an opportunity to make things easier to review?
```suggestion
        $container->setDefinition(
            sprintf('doctrine.dbal.%s_connection.event_manager', $name), 
            new ChildDefinition('doctrine.dbal.connection.event_manager')
        );
```

File: DependencyInjection/DoctrineExtension.php [hunk 9]
```diff
@@ -168,13 +162,13 @@ protected function loadDbalConnection($name, array $connection, ContainerBuilder
...
 
         // connection
         $options = $this->getConnectionOptions($connection);
 
         $def = $container
-            ->setDefinition(sprintf('doctrine.dbal.%s_connection', $name), new $definitionClassname('doctrine.dbal.connection'))
+            ->setDefinition(sprintf('doctrine.dbal.%s_connection', $name), new ChildDefinition('doctrine.dbal.connection'))
```
Comment (greg0ire) [inline DependencyInjection/DoctrineExtension.php]:
```suggestion
            ->setDefinition(
                sprintf('doctrine.dbal.%s_connection', $name),
                new ChildDefinition('doctrine.dbal.connection')
            )
```

File: DependencyInjection/DoctrineExtension.php [hunk 10]
```diff
@@ -407,6 +381,7 @@ protected function ormLoad(array $config, ContainerBuilder $container)
          * This is replaced with a true locator by ServiceRepositoryCompilerPass.
          * This makes that pass technically optional (good for tests).
          */
+        // @todo cleanup
```
Comment (greg0ire) [inline DependencyInjection/DoctrineExtension.php]: todo before merging?

File: DependencyInjection/DoctrineExtension.php [hunk 11]
```diff
@@ -423,8 +398,7 @@ protected function ormLoad(array $config, ContainerBuilder $container)
      */
     protected function loadOrmEntityManager(array $entityManager, ContainerBuilder $container)
     {
-        $definitionClassname = $this->getDefinitionClassname();
-        $ormConfigDef        = $container->setDefinition(sprintf('doctrine.orm.%s_configuration', $entityManager['name']), new $definitionClassname('doctrine.orm.configuration'));
+        $ormConfigDef = $container->setDefinition(sprintf('doctrine.orm.%s_configuration', $entityManager['name']), new ChildDefinition('doctrine.orm.configuration'));
```
Comment (greg0ire) [inline DependencyInjection/DoctrineExtension.php]:
```suggestion
        $ormConfigDef = $container->setDefinition(
            sprintf('doctrine.orm.%s_configuration', $entityManager['name']),
            new ChildDefinition('doctrine.orm.configuration')
        );
```

File: DependencyInjection/DoctrineExtension.php [hunk 12]
```diff
@@ -447,30 +421,18 @@ protected function loadOrmEntityManager(array $entityManager, ContainerBuilder $
             'setAutoGenerateProxyClasses' => '%doctrine.orm.auto_generate_proxy_classes%',
             'setClassMetadataFactoryName' => $entityManager['class_metadata_factory_name'],
             'setDefaultRepositoryClassName' => $entityManager['default_repository_class'],
+            'setNamingStrategy' => new Reference($entityManager['naming_strategy']),
+            'setQuoteStrategy' => new Reference($entityManager['quote_strategy']),
+            'setEntityListenerResolver' => new Reference(sprintf('doctrine.orm.%s_entity_listener_resolver', $entityManager['name'])),
```
Comment (greg0ire) [inline DependencyInjection/DoctrineExtension.php]:
```suggestion
            'setEntityListenerResolver' => new Reference(sprintf(
                'doctrine.orm.%s_entity_listener_resolver',
                $entityManager['name'])
            ),
```

File: DependencyInjection/DoctrineExtension.php [hunk 13]
```diff
@@ -447,30 +421,18 @@ protected function loadOrmEntityManager(array $entityManager, ContainerBuilder $
...
-        if (version_compare(Version::VERSION, '2.5.0-DEV') >= 0) {
-            $listenerId        = sprintf('doctrine.orm.%s_listeners.attach_entity_listeners', $entityManager['name']);
-            $listenerDef       = $container->setDefinition($listenerId, new Definition('%doctrine.orm.listeners.attach_entity_listeners.class%'));
-            $listenerTagParams = ['event' => 'loadClassMetadata'];
-            if (isset($entityManager['connection'])) {
-                $listenerTagParams['connection'] = $entityManager['connection'];
-            }
-            $listenerDef->addTag('doctrine.event_listener', $listenerTagParams);
+        $listenerId        = sprintf('doctrine.orm.%s_listeners.attach_entity_listeners', $entityManager['name']);
+        $listenerDef       = $container->setDefinition($listenerId, new Definition('%doctrine.orm.listeners.attach_entity_listeners.class%'));
```
Comment (greg0ire) [inline DependencyInjection/DoctrineExtension.php]:
```suggestion
        $listenerDef       = $container->setDefinition(
            $listenerId,
            new Definition('%doctrine.orm.listeners.attach_entity_listeners.class%')
        );
```

File: DependencyInjection/DoctrineExtension.php [hunk 14]
```diff
@@ -516,7 +478,7 @@ protected function loadOrmEntityManager(array $entityManager, ContainerBuilder $
 
         $managerConfiguratorName = sprintf('doctrine.orm.%s_manager_configurator', $entityManager['name']);
         $container
-            ->setDefinition($managerConfiguratorName, new $definitionClassname('doctrine.orm.manager_configurator.abstract'))
+            ->setDefinition($managerConfiguratorName, new ChildDefinition('doctrine.orm.manager_configurator.abstract'))
```
Comment (greg0ire) [inline DependencyInjection/DoctrineExtension.php]:
```suggestion
            ->setDefinition(
                $managerConfiguratorName,
                new ChildDefinition('doctrine.orm.manager_configurator.abstract')
            )
```

File: DependencyInjection/DoctrineExtension.php [hunk 15]
```diff
@@ -525,7 +487,7 @@ protected function loadOrmEntityManager(array $entityManager, ContainerBuilder $
         }
 
         $container
-            ->setDefinition(sprintf('doctrine.orm.%s_entity_manager', $entityManager['name']), new $definitionClassname('doctrine.orm.entity_manager.abstract'))
+            ->setDefinition(sprintf('doctrine.orm.%s_entity_manager', $entityManager['name']), new ChildDefinition('doctrine.orm.entity_manager.abstract'))
```
Comment (greg0ire) [inline DependencyInjection/DoctrineExtension.php]:
```suggestion
            ->setDefinition(
                sprintf('doctrine.orm.%s_entity_manager',
                $entityManager['name']), new ChildDefinition('doctrine.orm.entity_manager.abstract')
            )
```

File: Tests/ContainerTest.php [hunk 16]
```diff
@@ -48,14 +49,9 @@ public function testContainer()
...
-        if (interface_exists('Symfony\Component\PropertyInfo\PropertyInfoExtractorInterface') && class_exists('Symfony\Bridge\Doctrine\PropertyInfo\DoctrineExtractor')) {
-            if (! interface_exists(PropertyInitializableExtractorInterface::class)) {
-                $this->assertInstanceOf('Doctrine\Common\Persistence\Mapping\ClassMetadataFactory', $container->get('doctrine.orm.default_entity_manager.metadata_factory'));
-            }
-            $this->assertInstanceOf('Symfony\Bridge\Doctrine\PropertyInfo\DoctrineExtractor', $container->get('doctrine.orm.default_entity_manager.property_info_extractor'));
-        } else {
-            $this->assertFalse($container->has('doctrine.orm.default_entity_manager.metadata_factory'));
-            $this->assertFalse($container->has('doctrine.orm.default_entity_manager.property_info_extractor'));
+        if (! interface_exists(PropertyInitializableExtractorInterface::class)) {
+            $this->assertInstanceOf('Doctrine\Common\Persistence\Mapping\ClassMetadataFactory', $container->get('doctrine.orm.default_entity_manager.metadata_factory'));
```
Comment (greg0ire) [inline Tests/ContainerTest.php]:
```suggestion
            $this->assertInstanceOf(
                'Doctrine\Common\Persistence\Mapping\ClassMetadataFactory',
                $container->get('doctrine.orm.default_entity_manager.metadata_factory')
            );
```

File: Tests/ContainerTest.php [hunk 17]
```diff
@@ -48,14 +49,9 @@ public function testContainer()
...
         }
+        $this->assertInstanceOf('Symfony\Bridge\Doctrine\PropertyInfo\DoctrineExtractor', $container->get('doctrine.orm.default_entity_manager.property_info_extractor'));
```
Comment (greg0ire) [inline Tests/ContainerTest.php]:
```suggestion
        $this->assertInstanceOf(
            'Symfony\Bridge\Doctrine\PropertyInfo\DoctrineExtractor',
            $container->get('doctrine.orm.default_entity_manager.property_info_extractor')
        );
```

File: Tests/DependencyInjection/AbstractDoctrineExtensionTest.php [hunk 18]
```diff
@@ -817,14 +772,10 @@ public function testEntityListenerResolver()
         $container = $this->loadContainer('orm_entity_listener_resolver', ['YamlBundle'], new EntityListenerPass());
 
         $definition = $container->getDefinition('doctrine.orm.em1_configuration');
-        if (version_compare(Version::VERSION, '2.4.0-DEV') >= 0) {
-            $this->assertDICDefinitionMethodCallOnce($definition, 'setEntityListenerResolver', [new Reference('doctrine.orm.em1_entity_listener_resolver')]);
-        }
+        $this->assertDICDefinitionMethodCallOnce($definition, 'setEntityListenerResolver', [new Reference('doctrine.orm.em1_entity_listener_resolver')]);
```
Comment (greg0ire) [inline Tests/DependencyInjection/AbstractDoctrineExtensionTest.php]:
```suggestion
        $this->assertDICDefinitionMethodCallOnce(
            $definition, 'setEntityListenerResolver',
            [new Reference('doctrine.orm.em1_entity_listener_resolver')]
        );
```

File: Tests/DependencyInjection/AbstractDoctrineExtensionTest.php [hunk 19]
```diff
@@ -817,14 +772,10 @@ public function testEntityListenerResolver()
...
 
         $definition = $container->getDefinition('doctrine.orm.em2_configuration');
-        if (version_compare(Version::VERSION, '2.4.0-DEV') >= 0) {
-            $this->assertDICDefinitionMethodCallOnce($definition, 'setEntityListenerResolver', [new Reference('doctrine.orm.em2_entity_listener_resolver')]);
-        }
+        $this->assertDICDefinitionMethodCallOnce($definition, 'setEntityListenerResolver', [new Reference('doctrine.orm.em2_entity_listener_resolver')]);
```
Comment (greg0ire) [inline Tests/DependencyInjection/AbstractDoctrineExtensionTest.php]:
```suggestion
        $this->assertDICDefinitionMethodCallOnce(
            $definition, 'setEntityListenerResolver',
            [new Reference('doctrine.orm.em2_entity_listener_resolver')]
        );
```

**Comment by ostrolucky** [comment]
Not worth doing job of cs fixers manually

**Comment by greg0ire** [comment]
Is there a fixer for this?

**Comment by ostrolucky** [comment]
https://github.com/Symplify/CodingStandard#parameters-arguments-and-array-items-should-be-on-the-samestandalone-line-to-fit-line-length

File: DependencyInjection/DoctrineExtension.php (hunk 10 above)
Comment (alcaeus) [reply DependencyInjection/DoctrineExtension.php]: Ah, almost forgot about that. I'll update this 👍

**Comment by alcaeus** [comment]
@greg0ire I'll work on the todo, completely forgot about that, sorry.

As for CS, I'd like to avoid style changes in this PR. doctrine/coding-standard also doesn't have any line-length rules, so I'd like to direct discussion about that to https://github.com/doctrine/coding-standard so we can come up with consistent rules on that. Thanks!

File: DependencyInjection/DoctrineExtension.php (hunk 10 above)
Comment (alcaeus) [reply DependencyInjection/DoctrineExtension.php]: Done!

**Comment by greg0ire** [comment]
I'm AFK for a week so cannot create the cs discussion, especially since it seems to mean creating a PR, but feel free to
